src/vsr/replica.zig

```diff
@@ -2520,6 +2520,49 @@ pub fn Replica(
             return message.ref();
         }
 
+        /// Discards headers during a view change that may not be canonical, if the new primary did
+        /// not participate in any previous view changes that may have removed or changed these ops.
+        /// This is required for correctness and liveness, to ensure that:
+        /// 1. disconnected headers do not remain in place in lieu of gaps, so that
+        /// 2. gaps may be discarded by discard_uncommitted_headers().
+        fn discard_uncanonical_headers(self: *Self, view_normal_canonical: u64) void {
+            assert(self.replica_count > 1);
+            assert(self.status == .view_change);
+            assert(self.leader_index(self.view) == self.replica);
+            assert(self.do_view_change_quorum);
+            assert(!self.repair_timeout.ticking);
+            assert(self.journal.header_with_op(self.op) != null);
+            assert(self.view_normal <= view_normal_canonical);
+
+            if (self.view_normal == view_normal_canonical) return;
+
+            const remove_op_count = std.math.min(
+                // Do not reset any ops that we have already committed.
+                self.op - self.commit_min,
+                // The number of uncommitted ops cannot be more than the length of the pipeline.
+                // Do not reset any ops that we did not include in our do_view_change message.
+                config.pipeline_max,
+            );
+
+            assert(remove_op_count <= config.pipeline_max);
+            if (remove_op_count == 0) return;
+
+            const remove_op_min = self.op -| (remove_op_count - 1); // Inclusive.
```

Comment:
Is the saturating subtraction necessary?

- When `remove_op_count = self.op - self.commit_min`, `self.op - (remove_op_count - 1) = (self.op - remove_op_count) + 1 = commit_min + 1`.
- When `remove_op_count = config.pipeline_max`, `config.pipeline_max < self.op - self.commit_min < self.op`.

Reply:
This is what I have:

```zig
const remove_op_count = std.math.min(
    // Do not reset any ops that we have already committed.
    self.op - self.commit_min,
    // The number of uncommitted ops cannot be more than the length of the pipeline.
    // Do not reset any ops that we did not include in our do_view_change message.
    config.pipeline_max,
);

assert(remove_op_count <= config.pipeline_max);
if (remove_op_count == 0) return;

// Saturating subtraction is not necessary below, because we take the min() above:
// * When `remove_op_count` = `self.op - self.commit_min`,
//      `self.op - self.commit_min <= self.op`.
// * When `remove_op_count = config.pipeline_max`,
//      `config.pipeline_max < self.op - self.commit_min <= self.op`.
const remove_op_min = self.op - remove_op_count + 1; // Add 1 as `op_min` is inclusive.
```

However, I think this is harder to get wrong, as it also binds intention with respect to the replica's DVC:

```zig
const message = self.do_view_change_from_all_replicas[self.replica].?;
assert(message.header.view == self.view);
assert(message.header.op == self.op);

var op_min: ?u64 = null;
var op_count: usize = config.pipeline_max;

for (message_body_as_headers(message)) |header| {
    assert(op_min == null or op_min.? > header.op);

    if (header.op <= self.commit_min) continue;
    if (op_count == 0) continue; // Assert all subsequent header.op's against op_min.

    op_min = header.op;
    op_count -= 1;
}
```

Now we have a stronger guarantee that any header we discard here as potentially uncanonical, will be able to be repaired later from the replica's DVC, if it turns out to be canonical.

Reply:
Can we add
```
assert(message.header.commit == self.commit_min);
```
to the `message`'s assertions?

---

`op_min` is a tricky name, since it is about to become the new `replica.op` (which is the replica's _maximum_ op). What about `op_uncanonical_min` or `op_truncate`?

---

Does the `op_count` decrement track correctly when there are gaps in the DVC headers?

Reply:
Also, when the `if (op_count == 0)` branch is taken, doesn't that mean that more than `config.pipeline_max` ops will be removed? i.e. shouldn't it be

```zig
if (op_count == 0) {
    op_min = op_min.? + 1;
    continue
}
```